crates/edr_napi_core/src/subscription.rs

```diff
@@ -1,23 +1,102 @@
-use derive_where::derive_where;
-use edr_eth::B256;
-use edr_provider::{time::CurrentTime, ProviderSpec, SubscriptionEvent};
+use std::sync::Arc;
+
+use edr_eth::{filter::LogOutput, B256, U256};
+use edr_evm::BlockAndTotalDifficulty;
 use napi::{
     threadsafe_function::{
         ErrorStrategy, ThreadSafeCallContext, ThreadsafeFunction, ThreadsafeFunctionCallMode,
     },
-    JsFunction,
+    JsFunction, JsUnknown,
 };
 
-#[derive_where(Clone)]
-pub struct Callback<ChainSpecT: ProviderSpec<CurrentTime> + 'static> {
-    inner: ThreadsafeFunction<SubscriptionEvent<ChainSpecT>, ErrorStrategy::Fatal>,
+/// A chain-agnostic version of [`edr_provider::SubscriptionEvent`].
+pub struct SubscriptionEvent {
+    pub filter_id: U256,
+    pub result: SubscriptionEventData,
+}
+
+impl SubscriptionEvent {
+    pub fn new<BlockT, RpcBlockT, SignedTransactionT>(
+        event: edr_provider::SubscriptionEvent<BlockT, SignedTransactionT>,
+    ) -> Self
+    where
+        BlockT: ?Sized + 'static,
+        RpcBlockT:
+            From<BlockAndTotalDifficulty<Arc<BlockT>, SignedTransactionT>> + serde::Serialize,
+        SignedTransactionT: 'static,
+    {
+        let edr_provider::SubscriptionEvent { filter_id, result } = event;
+
+        Self {
+            filter_id,
+            result: SubscriptionEventData::new::<_, RpcBlockT, _>(result),
+        }
+    }
+}
+
+/// Type alias for a function that converts a value to a JavaScript value.
```

Comment:
```suggestion
/// Type alias for a closure that produces a JavaScript value given a `napi::Env`.
```

Reply:
Addressed in https://github.com/NomicFoundation/edr/pull/1040/commits/375b694bb6a684d25adfc612eeb5c3c8b1eb1b04